### PR TITLE
create soft links from overlays

### PR DESCRIPTION
**Pull request by mslacken**
Add the the function softlink to the template rendering, so that
{{ softlink: "/bin/bash" }} for the file `/bin/run.ww` will create
a soft link to `/bin/bash`
It's also possible to "import" soft links from the host, e.g. the
template
{{ ImportLink "/etc/localtime" }}
set the timezone from the host.

### CODE REVIEW

File: overlays/wwinit/rootfs/etc/localtime.ww [hunk 1]
```diff
@@ -0,0 +1 @@
+{{ ImportLink "/etc/localtime" }}
```
Comment (anderbubble) [inline overlays/wwinit/rootfs/etc/localtime.ww:1]: Can you move this to a new overlay? Something like `localtime` maybe, unless there's another one that makes sense.

File: internal/pkg/overlay/funcmap.go [hunk 2]
```diff
@@ -122,3 +123,12 @@ func createIgnitionJson(node *node.NodeConf) string {
...
 }
+
+func importSoftlink(lnk string) string {
+	path, err := filepath.EvalSymlinks(lnk)
+	if err != nil {
+		return "abort"
+	}
+	wwlog.Debug("importing softlink pointing to: %s", path)
+	return fmt.Sprintf("{{ /* softlink \"%s\" */ }}", path)
+}
```
Comment (anderbubble) [inline internal/pkg/overlay/funcmap.go:134]: Could this be replaced with a value that reads the target of a symlink, which could then be passed to the other symlink function? That justifies both functions (by using them both in the example localtime use case) and is more flexible for other templates later.

File: internal/pkg/overlay/overlay.go [hunk 3]
```diff
@@ -238,12 +238,17 @@ func BuildOverlayIndir(nodeData node.NodeConf, overlayNames []string, outputDir
...
 					for fileScanner.Scan() {
 						line := fileScanner.Text()
-						filenameFromTemplate := reg.FindAllStringSubmatch(line, -1)
-						if len(filenameFromTemplate) != 0 {
+						filenameFromTemplate := regFile.FindAllStringSubmatch(line, -1)
+						softlinkFromTemplate := regLink.FindAllStringSubmatch(line, -1)
+						if len(softlinkFromTemplate) != 0 {
+							wwlog.Debug("Creating soft link %s -> %s", destFileName, softlinkFromTemplate[0][1])
+							return os.Symlink(softlinkFromTemplate[0][1], path.Join(outputDir, destFileName))
+						} else if len(filenameFromTemplate) != 0 {
```
Comment (anderbubble) [inline internal/pkg/overlay/overlay.go:251]: This implicitly uses the symlink value if it exists, ignoring any `file` directive. I think specifying both `symlink` and `file` should be an error. Frankly, if there's a `symlink` directive and there's anything else generated that _isn't_ the `symlink` directive, I think that should probably be an error.

**Review by anderbubble — CHANGES_REQUESTED** [review]
I _finally_ think I understand what this is for.

I left some comments in the requested changes. Let me know what you think.

**Comment by anderbubble** [comment]
Rebased to main.

**Comment by mslacken** [comment]
rebased

**Comment by anderbubble** [comment]
Rebased to main.

**Review by anderbubble — APPROVED** [review]
@mslacken I added tests for the "localtime" overlay and made it configurable. If you approve, I'll go ahead and merge.

**Comment by anderbubble** [comment]
Merging as discussed in 18 December 2024 TSC meeting.

**Comment by anderbubble** [comment]
Closed #1060

**Pull request merged**